xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3446,11 +3457,148 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+            }
+            SortedMap<Integer, Map<String, String[]>> objectMap = result.get(classReference);
+            if (objectMap == null) {
+                objectMap = new TreeMap<>();
+                result.put(classReference, objectMap);
+            }
+            // Get the property from the right object #objectNumber of type 'objectName'; create it if they don't exist
+            Map<String, String[]> object = objectMap.get(classNumber);
+            if (object == null) {
+                object = new HashMap<>();
+                objectMap.put(classNumber, object);
+            }
+            object.put(classPropertyName, parameter.getValue());
+        }
+        return result;
+    }
+
+    /**
+     * Create and/or update objects in a document given a list of HTTP parameters of the form
+     * '<spacename>.<classname>_<number>_<propertyname>'. If the object already exists, the field is replace by the
+     * given value. If the object doesn't exist in the document, it is created then the property <propertyname> is
+     * initialized with the given value. An object is only created if the given '<number>' is 'one-more' than the
+     * existing number of objects. For example, if the document already has 2 objects of type 'Space.Class', then it
+     * will create a new object only with 'Space.Class_2_prop=something'. Every other parameter like
+     * 'Space.Class_42_prop=foobar' for example, will be ignore.
+     * 
+     * @param eform is form information that contains all the query parameters
+     * @param context
+     * @throws XWikiException
+     * @since 7.0RC1
+     */
+    public void readObjectsFromFormUpdateOrCreate(EditForm eform, XWikiContext context) throws XWikiException
+    {
+        Map<DocumentReference, SortedMap<Integer, Map<String, String[]>>> fromRequest =
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<DocumentReference, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest
+            .entrySet()) {
+            DocumentReference requestClassReference = requestClassEntries.getKey();
+            SortedMap<Integer, Map<String, String[]>> requestObjectMap = requestClassEntries.getValue();
+            List<BaseObject> newObjects = new ArrayList<>();
+            for (Entry<Integer, Map<String, String[]>> requestObjectEntry : requestObjectMap.entrySet()) {
+                Integer requestObjectNumber = requestObjectEntry.getKey();
+                Map<String, String[]> requestObjectPropertyMap = requestObjectEntry.getValue();
+                BaseObject oldObject = getXObject(requestClassReference, requestObjectNumber);
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
+                    if (requestObjectPropertyMap != null
+                        && requestObjectNumber == getXObjectSize(requestClassReference)) {
```

Comment:
My choice is based on the answer to this question: "what will you do if there is no object in the wiki page and your client gives you something like `XWiki.XWikiRights_42_levels`?".  You should create 42 objects only for one useful object?  That's the reason I made this choice.  I kind of force the developer to know what he's doing.  It's only a limitation, may be improved in a later PR?